text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+trait RngExt {
+    fn gen<T, D>(&mut self, data: D) -> T where T: Random<D> { ... }
+
+    /// Return an iterator that generates an infinite sequence of random values of type `T`.
+    fn gen_iter<T, D>(self, data: D) -> Generator<T, D, Self> where T: Random<D> { ... }
+
+    /// Randomize the order of the elements of `values`
+    fn shuffle<T>(&mut self, values: &mut [T]) { ... }
+    /// Select one value uniformly at random from the slice.
+    #[unstable]
+    fn choose<'a>(&mut self, values: &[T]) -> Option<&'a T> { ... }
+    /// Select `n` values uniformly at random from the iterator.
+    fn sample<I: Iterator>(&mut self, n: uint, values: I) -> Vec<I::Item> { ... }
+}
+
+impl<T: Rng> RngExt for T {}
+```
+
+As today, the relationship between output of the methods of `Rng`, and
+between them and `io::Reader` is unspecified.
+
+### The `Random` trait
+
+The change to `Random` (currently `Rand`) is motivated by allowing
+`gen` and `gen_range` to be unified, what was previously
+`rng.gen_range(a, b)` can become `rng.gen(a..b)`. However, this makes
+the non-ranged generation slightly more complicated/"magic
+incantation"-y: `rng.gen(..)` (this relies on having explicit syntax
+for `FullRange`, per [RFC 702]).
+
+[RFC 702]: https://github.com/rust-lang/rfcs/pull/702
+
+The convention for using `FullRange` for unrestricted generation may
+seem peculiar for types that don't have any concept of ranges,
+e.g. creating a randomly seeded RNG, but the consistency and
+"vagueness" of `..` likely means this is OK.
+
+(NB. general discussion, non-normative) This approach allows us to
+naturally leverage any "range combinators" used elsewhere,
+e.g. inclusive endpoints, and so remove the need for the `Closed01`
+and `Open01` types. It also allows us to implement `Random` for
+e.g. `Vec`, say, `rng.gen::<Vec<u8>>(10..20)` would create a vector of
+some length between 10 and 20 and fill it with `u8`s.
+
+#### `#[derive]`
+
+The deriving mode will be removed from the compiler, or at least,
+disabled on the stable channel; it can be rarely used since most data
+types have some restriction on their values and/or non-trivial
+relationships between them, which a derived `Random` implementation
+can do nothing to preserve.
+
+#### Alternatives
+
+Call it `Rand` as it is today.
+
+Not introduce the `Data` parameter. This would be simpler for this
+trait, but require duplication elsewhere.
+
+Have either `Self` or `Data` be associated types. This would require
+more proxy types since e.g. the following is illegal:
+
+```rust
+impl Random for u8 {
+    type Data = FullRange;
+    // ...
+}
+impl Random for u8 {
+    type Data = Range<u8>;
+    // ...
+}
+```
+
+### The `Rng` trait
+
+[the `Rng` trait]: #the-rng-trait
+
+This trait represent random number generators, it is mostly provided
+for expressing intent, and for optimisation purposes over a plain
+`Reader`.
+
+It is expected and recommended that types implementing `Rng` generates
```

Comment:
This is referring to the `Reader` implementation? So they are *not* intended to generate random `u8`s?